### ISSUE TITLE
ColumnEncodingUtility: Crash when reserved word is in column name

### ISSUE THREAD

**Issue by lupujo** [awslabs/amazon-redshift-utils#329]
The script always fails with the following error if one of the column names is called 'time' (perhaps in more cases as well):

```
-- [3033] [3033] Running insert into public."table_$mig"  select * from public."table" order by id,time,action;
Traceback (most recent call last):
  File "./analyze-schema-compression.py", line 420, in run_commands
    cursor.execute(c)
  File "/usr/local/lib/python2.7/site-packages/pg8000/core.py", line 884, in execute
    self._c.execute(self, operation, args)
  File "/usr/local/lib/python2.7/site-packages/pg8000/core.py", line 1839, in execute
    self.handle_messages(cursor)
  File "/usr/local/lib/python2.7/site-packages/pg8000/core.py", line 1978, in handle_messages
    raise self.error
ProgrammingError: (u'ERROR', u'42601', u'syntax error at or near ","', u'141', u'/home/ec2-user/padb/src/pg/src/backend/parser/parser_scan.l', u'707', u'yyerror')

```

The reason is that some columns (like 'time') require double quotes due to them being reserved words.
The following patch seems to fix the problem by changing line 785 from:
`insert = "%s order by %s;" % (insert, ",".join(table_sortkeys))`
To:
`insert = "%s order by \"%s\";" % (insert, ",".join(table_sortkeys).replace(',','\",\"'))`

**Comment by patbl** [comment]
This seems still to be a problem. My `time` column is still unescaped in `c` when this line is run:

https://github.com/awslabs/amazon-redshift-utils/blob/5b4f8e9253981454d2305e322d2664d592306c75/src/ColumnEncodingUtility/analyze-schema-compression.py#L436

**Comment by IanMeyers** [comment]
Looking

**Comment by IanMeyers** [comment]
Can you provide some log output? When I try to reproduce, it all looks OK:

```
- [29186] [29186] Running set statement_timeout = '1200000'
-- [29186] Success.
-- [29186] [29186] Running set application_name to 'ColumnEncodingUtility-v.9.3.4'
-- [29186] Success.
-- [29186] Connected to localhost:5439:master as master
-- [29186] Analyzing Table 'test_analysis' for Columnar Encoding Optimisations with 1 Threads...
-- [29186] Analyzing 1 table(s) which contain allocated data blocks
-- [29186] Table public.test_analysis contains 2 unoptimised columns
-- [29186] Using Force Override Option
-- [29186] Analyzing Table 'public.test_analysis'
-- [29186] Column Encoding will be modified for public.test_analysis
-- [29186] No encoding modifications run for public.test_analysis
begin;
lock table public."test_analysis";
create table public."test_analysis_$mig"(
"id" bigint  NOT NULL encode raw
,"column_a" varchar(100)   encode raw
,"time" timestamp    encode RAW
)
 SORTKEY(time)
 ;
alter table public."test_analysis_$mig" add primary key (id);
alter table public."test_analysis_$mig" owner to master;
insert into public."test_analysis_$mig"  select * from public."test_analysis" order by "time";
analyze public."test_analysis_$mig";
alter table public."test_analysis" rename to "test_analysis_20181127_sKQpLnRcxf_$old";
alter table public."test_analysis_$mig" rename to "test_analysis";
commit;
-- [29186] Performed modification of 1 tables
-- [29186] Processing Complete
```

**Comment by patbl** [comment]
I removed or renamed a bunch of stuff, since I don't know whether my company wants its database schema to be public.

```
./analyze-schema-compression.py \
  --db db_name \
  --db-user db_user \
  --db-host lol.us-east-1.redshift.amazonaws.com \
  --analyze-table my_table \
  --debug true \
  --do-execute true
[...]
-- [25298] [25298] Running insert into public."my_table_$mig" ([...], time, [...]) select [...], time, [...] from public."my_table";
Traceback (most recent call last):
  File "./analyze-schema-compression.py", line 436, in run_commands
    cursor.execute(c)
  File "/usr/local/lib/python2.7/dist-packages/pg8000/core.py", line 892, in execute
    self._c.execute(self, operation, args)
  File "/usr/local/lib/python2.7/dist-packages/pg8000/core.py", line 1849, in execute
    self.handle_messages(cursor)
  File "/usr/local/lib/python2.7/dist-packages/pg8000/core.py", line 1988, in handle_messages
    raise self.error
ProgrammingError: (u'ERROR', u'42601', u'syntax error at or near ","', u'432', u'/home/ec2-user/padb/src/pg/src/backend/parser/parser_scan.l', u'709', u'yyerror')

Error running statements: [...]
Error in worker thread: return code 1. Exiting.

begin;
lock table public."my_table";
create table public."my_table_$mig"(
[...]
,"time" timestamp    encode zstd
[...]
)
;
alter table public."my_table_$mig" add primary key (id);
alter table public."my_table_$mig" owner to lol;
insert into public."my_table_$mig" ([...] time [...]) select [...], time, [...] from public."my_table";
```

**Comment by IanMeyers** [comment]
Sorry it's taken me so long to address this. I believe this is fixed with 9b4a50e:

```
8c85904b1c99:amazon-redshift-utils meyersi$ docker run --net host --rm -it --env-file .env amazon-redshift-utils column-encoding
Running column-encoding utility
-- [8] [8] Running set statement_timeout = '1200000'
-- [8] Success.
-- [8] [8] Running set application_name to 'ColumnEncodingUtility-v.9.3.4'
-- [8] Success.
-- [8] Connected to host.docker.internal:5439:master as master
-- [8] Analyzing Table 'issue_329' for Columnar Encoding Optimisations with 1 Threads...
-- [8] Analyzing 1 table(s) which contain allocated data blocks
-- [8] Table public.issue_329 contains 0 unoptimised columns
-- [8] Using Force Override Option
-- [8] Analyzing Table 'public.issue_329'
-- [8] Column Encoding will be modified for public.issue_329
-- [8] No encoding modifications run for public.issue_329
begin;
lock table public."issue_329";
create table public."issue_329_$mig"(
"id" bigint identity (1, 1)  encode raw
,"column_a" varchar(10)  NOT NULL encode raw
,"column_b" integer   encode raw
,"column_c" date default ('now'::text)::date  encode raw
,"time" timestamp    encode raw
)
;
alter table public."issue_329_$mig" owner to master;
insert into public."issue_329_$mig" ("column_a", "column_b", "column_c", "time") select "column_a", "column_b", "column_c", "time" from public."issue_329";
analyze public."issue_329_$mig";
alter table public."issue_329" rename to "issue_329_20181212_EN5k6cJJgd_$old";
alter table public."issue_329_$mig" rename to "issue_329";
commit;
-- [8] Performed modification of 1 tables
-- [8] Processing Complete
Done
```

You can see that all column names are quoted in the generated migration script.

**Comment by saeedma8** [comment]
[Amazon Redshift Column Encoding Utility](https://github.com/awslabs/amazon-redshift-utils/tree/master/src/ColumnEncodingUtility#amazon-redshift-column-encoding-utility) is now deprecated. Instead, please use Redshift's Automatic Table Optimization features through:

```
ALTER TABLE table_name ALTER SORTKEY AUTO;
ALTER TABLE table_name ALTER DISTSTYLE AUTO;
```

https://docs.aws.amazon.com/redshift/latest/dg/t_Creating_tables.html
